BUGFIX: Fix warnings from GCC and Clang

jpihl opened the pull request:
Hello,

This PR is basically a duplicate of #20, but it seems the work there has stalled a bit. I tried looking at your suggestions to @dan-42 and address the warnings accordingly.

Let me know if you think something needs to be handled differently.

jpihl commented:
Hi @oktal,

I know you are busy, but it would be great to get this PR merged. The warnings seems to be problematic for many people.

markusgau commented:
Could this be rebase once more mabye? There are still a lot of issues when building for ARM64 with -pedantic and -Wextra enabled.
I think chances are that this might get merged then. The new maintainer seems much more responsive...

dennisjenkins75 commented on examples/hello_server.cc:
```diff
@@ -1,33 +1,38 @@
 
     HTTP_PROTOTYPE(HelloHandler)
 
-    void onRequest(const Http::Request& request, Http::ResponseWriter response) {
```
Please do not reformat the code when making changes unrelated to formatting.

Its fine that you "(void)request;", but there was no need to split line 17 into 4 lines. 

I am heavily inclined to NACK your PR as-is, as it introduces many unnecessary diffs.

jpihl replied:
Sorry about that. The initial PR was a bit cleaner and I may have been a bit careless during this second "update". Give me a few minutes and I'll fix the worst parts.

jpihl commented:
@dennisjenkins75 I tried cleaning up the PR a bit. I've used the `UNUSED` macro in favor of omitting the names of unused parameters. There's a lot of tailing white space which have been removed but my life is too short for reverting that - I hope you can see past it.

dennisjenkins75 commented on examples/hello_server.cc:
```diff
@@ -15,19 +15,24 @@ class HelloHandler : public Http::Handler {
+    std::cout << "Starting server. Test with the following command: "
+              << "\"curl http://127.0.0.1:" << addr.port() << "\"\n";
 
+    std::cout << "Press Enter to Exit" << '\n';
```
Why did you change the runtime behavior of this test program?

jpihl replied:
This change has nothing to do with the purpose of this PR. I'll revert it if you like.

I changed the it so that it's possible to stop the execution, this makes it possible to use tool like valgrind to check for memory leaks.

dennisjenkins75 replied:
Please revert this change.

dennisjenkins75 commented on examples/hello_server.cc:
```diff
@@ -15,19 +15,24 @@ class HelloHandler : public Http::Handler {
 
     void onRequest(const Http::Request& request, Http::ResponseWriter response) {
-        response.send(Http::Code::Ok, "Hello World");
+        (void) request;
```
Did you want to change this to "UNUSED(request)" as well?

jpihl replied:
Sure. I used the void cast here because it's an example and I thought the `UNUSED` macro might confuse new users.

dennisjenkins75 replied:
Well, the UNUSED macro is used throughout the rest of the code, and imho, it is kinda self-explanatory.  Please change this to use the macro.  My main goal here is to remove the C-style cast.

dennisjenkins75 commented on include/pistache/client.h:
```diff
@@ -136,8 +136,8 @@ struct Connection : public std::enable_shared_from_this<Connection> {
 
-struct ConnectionPool {
-
+class ConnectionPool {
```
Why did you make this change?  What bug/problem does it address?  A class w/ all public methods/member vars is the same as a struct, so this change should be a no-op... but why make it?

jpihl replied:
The reason for this change is that ConnectionPool is forward declared as a class - which gives a warning. Alternatively I can revert this and forward declare it as a struct instead.

dennisjenkins75 replied:
Ah, I see.  Either direction (all struct or all class) is fine then.

dennisjenkins75 commented on include/pistache/common.h:
```diff
@@ -37,7 +37,7 @@
     [&]() { \
         auto ret = __VA_ARGS__; \
-        if (ret < 0) { \
+        if ((ssize_t)ret < 0) { \
```
Please use "static_cast<ssize_t>(ret)" instead of the older C-style casting.  Old-style casts are flagged as warnings when compiled with "-Weffc++".

jpihl replied:
I just tried removing the cast all together and I don't see any warnings with neither g++ or clang - I guess that's a better fix.

dennisjenkins75 commented on include/pistache/http.h:
```diff
@@ -549,7 +547,7 @@ class ResponseWriter : public Response {
        return &buf_;
     }
 
-    DynamicStreamBuf *rdbuf([[maybe_unused]] DynamicStreamBuf* other) {
```
Why are you removing the "[[maybe_unused]]"?  It was added only 28 days ago with PR #259.

jpihl replied:
I got warning for this. And then I read this comment also added in that PR:
https://github.com/oktal/pistache/pull/259/files#diff-5a4ac07535107245765a4a8d3f822a83R52

dennisjenkins75 replied:
I do not want to require c++17 any time soon.  So for the sake of consistency, removing this is fine.  Please find the other references are remove them as well (there is only one other, in that comment).

dennisjenkins75 commented on include/pistache/http.h:
```diff
@@ -613,31 +611,35 @@ namespace Private {
         Message *message;
     };
 
-    struct RequestLineStep : public Step {
```
Why are you changing the struct to a class?  What warning (and in what compiler version) is this change addressing?

jpihl replied:
This is addressing the same issue as the `ConnectionPool`.

dennisjenkins75 commented on src/common/http_defs.cc:
```diff
@@ -62,9 +62,10 @@ CacheDirective::delta() const
             return std::chrono::seconds(data.minFresh);
+        default:
+            throw std::domain_error("Invalid operation on cache directive");
+            return std::chrono::seconds();
```
This return will never execute.  Move it outside of the switch statement?  It won't execute their either...  I'm surprised that the compiler is not throwing a warning about missing a return statement.  Must be smart enough to realize that all cases are covered and that all will exit the method before the normal exit at the bottom.

I'm ambivalent on this one.  It just looks weird I guess.

jpihl replied:
You are right, it looks weird. I just tried removing the return statement and I no warning appeared. I'll remove the return statement.

dennisjenkins75 commented on src/common/os.cc:
```diff
@@ -19,7 +19,7 @@ using namespace std;
 namespace Pistache {
 
-int hardware_concurrency() {
+uint hardware_concurrency() {
```
https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency

std::thread::hardware_concurrency() returns an int.  I agree that it _should be_ an unsigned int, but the c++11 standard has it signed.

In fact, consider removing the entire body of this method and just "return std::thread::hardware_concurrency();".

jpihl replied:
Maybe I'm missing something, but in the link it's specified that `std::thread::hardware_concurrency()` returns an unsigned int?

I'll change the implementation to use `std::thread::hardware_concurrency()`.

dennisjenkins75 commented on src/common/reactor.cc:
```diff
@@ -55,8 +55,8 @@ struct Reactor::Impl {
  */
-struct SyncImpl : public Reactor::Impl {
-
+class SyncImpl : public Reactor::Impl {
```
Why change struct to class here?

jpihl replied:
Again same as `ConnectionPool`.

dennisjenkins75 commented on src/common/stream.cc:
```diff
@@ -71,7 +71,7 @@ DynamicStreamBuf::reserve(size_t size)
 bool
 StreamCursor::advance(size_t count) {
-    if (count > buf->in_avail())
+    if ((ssize_t)count > buf->in_avail())
```
Please use a C++ static_cast instead of a C-Style cast.

dennisjenkins75 commented on src/server/listener.cc:
```diff
@@ -220,7 +222,7 @@ Listener::run() {
                     if (event.flags.hasFlag(Polling::NotifyOn::Read)) {
                         auto fd = event.tag.value();
-                        if (fd == listen_fd)
+                        if ((ssize_t)fd == listen_fd)
```
Listener::listen_fd is an int (as it should be for file descriptors).  Just replace lines 224,255 with:

if (static_cast<int>(event.tag.value()) == listen_fd)

Unless "value()" already returns an int, then omit the static_cast.

jpihl replied:
`Tag::value()` returns a `uint64_t`, so the cast is necessary.

jpihl commented:
I don't currently have access to GCC-7.3.0 nor Clang-6.0.1.

dennisjenkins75 commented on examples/http_server.cc:
```diff
@@ -140,6 +140,7 @@ class MyHandler : public Http::Handler {
     }
 
     void onTimeout(const Http::Request& req, Http::ResponseWriter response) {
+        (void) req;
```
Please remove the C-style cast.  Replace w/ UNUSED() macro or c++ style cast.

jpihl commented:
I hope my latest changes has addressed all your comments. I've tested the compilation of this with the following compilers:
g++ (Ubuntu 7.3.0-16ubuntu3) 7.3.0
clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
